### ISSUE TITLE
`local_sum_make_vector` rewrite can introduce forbidden float64 operations at the graph level 

### ISSUE THREAD

**Issue by tvwenger** [pymc-devs/pytensor#653]
### Describe the issue:

This may be related to https://github.com/pymc-devs/pymc/issues/6779

**Description:**

When creating a model with `floatX="float32"` that includes ~~a Dirichlet distribution~~ a single distribution, the `floatX` assignment is respected. When creating a model with ~~a Dirichlet distribution as well as another distribution~~ two distributions, however, the `floatX` assignment is NOT respected, but only upon sampling. This is a weird bug.

**Expected Behavior**

The model should respect `floatX` in all cases.

**Actual Behavior**

When the model includes ~~a Dirichlet distribution and then ANY other distribution~~ two distributions, the graph includes `float64` despite the request that `floatX="float32"`.

**Minimum Working Example**

In the following MWE, I create four models. The first has one Dirichlet distribution, the second has one Normal distribution, and the remaining two include a Dirichlet distribution and then either a Normal or HalfCauchy distribution. 

The first two models sample without issue, and `floatX` is respected.

The second and third models raise `float64` errors *during sampling*. The error appears *after* `model.point_logps()`, which is what was all that was being checked in https://github.com/pymc-devs/pymc/issues/6779

The output (with truncated error messages) is appended below:

<details>

```shell
pytensor version: 2.18.6
pymc version: 5.10.4
pytensor.config.floatX =  float64

test_dirichlet
pytensor.config.floatX =  float32
foo float32
{'foo': -1.5}
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [foo]
Sampling 4 chains for 1_000 tune and 1_000 draw iterations (4_000 + 4_000 draws total) took 2 seconds.ns, 0 divergences]

test_normal
pytensor.config.floatX =  float32
foo float32
{'foo': -0.92}
Auto-assigning NUTS sampler...
Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [foo]
Sampling 4 chains for 1_000 tune and 1_000 draw iterations (4_000 + 4_000 draws total) took 2 seconds.ns, 0 divergences]

test_dirichlet_normal
pytensor.config.floatX =  float32
foo float32
bar float32
{'foo': -1.5, 'bar': -0.92}
ERROR (pytensor.graph.rewriting.basic): Rewrite failure due to: local_sum_make_vector
ERROR (pytensor.graph.rewriting.basic): node: Sum{axes=None}(MakeVector{dtype='float32'}.0)
ERROR (pytensor.graph.rewriting.basic): TRACEBACK:
ERROR (pytensor.graph.rewriting.basic): Traceback (most recent call last):
 ...
Exception: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [foo, bar]
Sampling 4 chains for 1_000 tune and 1_000 draw iterations (4_000 + 4_000 draws total) took 3 seconds.ns, 0 divergences]

test_dirichlet_halfcauchy
pytensor.config.floatX =  float32
foo float32
bar float32
{'foo': -1.5, 'bar': -1.14}
ERROR (pytensor.graph.rewriting.basic): Rewrite failure due to: local_sum_make_vector
ERROR (pytensor.graph.rewriting.basic): node: Sum{axes=None}(MakeVector{dtype='float32'}.0)
ERROR (pytensor.graph.rewriting.basic): TRACEBACK:
ERROR (pytensor.graph.rewriting.basic): Traceback (most recent call last):
...
Exception: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.

Multiprocess sampling (4 chains in 4 jobs)
NUTS: [foo, bar]
Sampling 4 chains for 1_000 tune and 1_000 draw iterations (4_000 + 4_000 draws total) took 3 seconds.ns, 0 divergences]
```

</details>

Note that the output of `print(model.point_logps())` demonstrates that the error occurs after `model.point_logps()`. The error occurs during sampling.

### Reproduceable code example:

```python
import pytensor
import pytensor.tensor as pt
import pymc as pm

print("pytensor version:", pytensor.__version__)
print("pymc version:", pm.__version__)

print("pytensor.config.floatX = ", pytensor.config.floatX)
print()


def test_dirichlet():
    print("test_dirichlet")
    print("pytensor.config.floatX = ", pytensor.config.floatX)
    with pm.Model() as model:
        foo = pm.Dirichlet("foo", a=pt.ones(3))
        print(foo, foo.dtype)
    print(model.point_logps())
    with model:
        trace = pm.sample()
    print()


def test_normal():
    print("test_normal")
    print("pytensor.config.floatX = ", pytensor.config.floatX)
    with pm.Model() as model:
        foo = pm.Normal("foo", mu=0.0, sigma=1.0)
        print(foo, foo.dtype)
    print(model.point_logps())
    with model:
        trace = pm.sample()
    print()


def test_dirichlet_normal():
    print("test_dirichlet_normal")
    print("pytensor.config.floatX = ", pytensor.config.floatX)
    with pm.Model() as model:
        foo = pm.Dirichlet("foo", a=pt.ones(3))
        print(foo, foo.dtype)
        bar = pm.Normal("bar", mu=0.0, sigma=1.0)
        print(bar, bar.dtype)
    print(model.point_logps())
    with model:
        trace = pm.sample()
    print()


def test_dirichlet_halfcauchy():
    print("test_dirichlet_halfcauchy")
    print("pytensor.config.floatX = ", pytensor.config.floatX)
    with pm.Model() as model:
        foo = pm.Dirichlet("foo", a=pt.ones(3))
        print(foo, foo.dtype)
        bar = pm.HalfCauchy("bar", beta=1.0)
        print(bar, bar.dtype)
    print(model.point_logps())
    with model:
        trace = pm.sample()
    print()


with pytensor.config.change_flags(warn_float64="raise", floatX="float32"):
    test_dirichlet()

with pytensor.config.change_flags(warn_float64="raise", floatX="float32"):
    test_normal()

with pytensor.config.change_flags(warn_float64="raise", floatX="float32"):
    test_dirichlet_normal()

with pytensor.config.change_flags(warn_float64="raise", floatX="float32"):
    test_dirichlet_halfcauchy()
```

### Error message:

<details>

```shell
ERROR (pytensor.graph.rewriting.basic): Rewrite failure due to: local_sum_make_vector
ERROR (pytensor.graph.rewriting.basic): node: Sum{axes=None}(MakeVector{dtype='float32'}.0)
ERROR (pytensor.graph.rewriting.basic): TRACEBACK:
ERROR (pytensor.graph.rewriting.basic): Traceback (most recent call last):
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/graph/rewriting/basic.py", line 1922, in process_node
    replacements = node_rewriter.transform(fgraph, node)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/graph/rewriting/basic.py", line 1082, in transform
    return self.fn(fgraph, node)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/tensor/rewriting/basic.py", line 950, in local_sum_make_vector
    add(*[cast(value, acc_dtype) for value in elements]), out_dtype
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/tensor/rewriting/basic.py", line 950, in <listcomp>
    add(*[cast(value, acc_dtype) for value in elements]), out_dtype
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/tensor/basic.py", line 763, in cast
    return _cast_mapping[dtype_name](x)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/graph/op.py", line 295, in __call__
    node = self.make_node(*inputs, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/tensor/elemwise.py", line 484, in make_node
    outputs = [
              ^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/tensor/elemwise.py", line 485, in <listcomp>
    TensorType(dtype=dtype, shape=shape)()
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/graph/type.py", line 228, in __call__
    return utils.add_tag_trace(self.make_variable(name))
                               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/graph/type.py", line 200, in make_variable
    return self.variable_type(self, None, name=name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/twenger/miniconda3/envs/pymc/lib/python3.11/site-packages/pytensor/tensor/variable.py", line 900, in __init__
    raise Exception(msg)
Exception: You are creating a TensorVariable with float64 dtype. You requested an action via the PyTensor flag warn_float64={ignore,warn,raise,pdb}.
```

</details>

### PyMC version information:

<details>
pytensor version: 2.18.6
pymc version: 5.10.4
</details>

### Context for the issue:

Models that include a Dirichlet distribution as well as any other distribution cannot use `float32`.

**Comment by ricardoV94** [comment]
The accumulator dtype shouldn't matter. That's internal and not what it's triggering the float64 check. Otherwise you couldn't ever sum inputs in float32.

The problem is the rewrite that removes the CAReduce is not watching out for this.

**Comment by ricardoV94** [comment]
Here is a minimal reproducible example:

```python
import pytensor
import pytensor.tensor as pt

with pytensor.config.change_flags(warn_float64="raise", floatX="float32"):
    x1,x2,x3 = pt.scalars("x1","x2","x3")
    out = pt.sum([x1, x2, x3], acc_dtype="float32")
    out.eval({x1:0.0, x2:1.0, x3:2.0})  # Fine

with pytensor.config.change_flags(warn_float64="raise", floatX="float32"):
    x1,x2,x3 = pt.scalars("x1","x2","x3")
    out = pt.sum([x1, x2, x3])
    out.eval({x1:0.0, x2:1.0, x3:2.0})
```

The problem is the rewrite exposes the internal `acc_dtype` when it replaces the `CAReduce` by an `Add`:

https://github.com/pymc-devs/pytensor/blob/8aeda39b408656d484923559777e876ae15f8687/pytensor/tensor/rewriting/basic.py#L948-L951

We could restrict the rewrite to only apply in cases where the internal acc_dtype doesn't have higher precision than the input/output dtypes

**Comment by tvwenger** [comment]
Aha, I see now. Thanks for the clarification @ricardoV94 

If you point me in the right direction, I can work on a PR.

**Comment by tvwenger** [comment]
Well, actually I couldn't quite figure out what you meant. Instead, I noticed that the function `local_sum_make_vector` is getting `acc_dtype` from the node, which is ultimately being set via the logic I posted earlier in `pytensor/tensor/elemwise.py`. In the linked PR, I simply changed the default behavior when `acc_dtype` is not specified so that instead of upcasting from the input `dtype` it defaults to `config.floatX`. I tested the MWE you posted and this simple change fixes the problem.

**Comment by ricardoV94** [comment]
What I was saying is that it may be more reasonable to opt out of the rewrite than trying to change the default internals of CAReduce

**Comment by tvwenger** [comment]
@ricardoV94 OK, I finally understand what you meant about skipping the rewrite entirely. I've implemented that fix in a new PR:  https://github.com/pymc-devs/pytensor/pull/656

**Comment by ricardoV94** [comment]
Closed via #659
